src/plugins/explore/public/components/query_panel/utils/type_detection/index.ts

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { LanguageType } from '../../types';
+
+export interface DetectionResult {
+  type: LanguageType;
+  confidence: number;
+  reason: string;
+  warnings: string[];
+}
+
+interface ScoreResult {
+  score: number;
+  reason: string;
+  warnings: string[];
+}
+
+export class QueryTypeDetector {
+  private pplStartPatterns: RegExp[];
+  private pplPipePatterns: RegExp[];
+  private nlStarters: RegExp[]; // nl: NATURAL LANGUAGE
+
+  constructor() {
+    this.pplStartPatterns = [
+      /^\s*source\b/i, // Matches "source" as a standalone word
+      /^\s*source\s*=/i, // Matches "source=" with optional spaces
+      /^\s*from\s+\w+/i, // Matches "from <word>"
+      /^\s*search\s+index\s*=/i, // Matches "search index="
+    ];
+    this.pplPipePatterns = [/\|\s*(where|filter|fields|sort|limit|stats|rename|eval)\b/i];
+    this.nlStarters = [
+      /^\s*(show|list|get|find|search|display|give|retrieve|fetch|tell)\b/i, // Common action verbs
+      /^\s*(what|how|when|where|why|which|who)\b/i, // Question starters
+      /^\s*(can|could|would|will|please|should|may|might|shall)\b/i, // Modal verbs
+      /^\s*(explain|describe|summarize|analyze|compare|calculate)\b/i, // Analytical verbs
+      /^\s*(is|are|was|were|do|does|did|has|have|had|am|be|being|been|the)\b/i, // Auxiliary verbs
+    ];
+  }
+
+  detect(query: string): DetectionResult {
+    const trimmedQuery = query.trim();
+    if (!trimmedQuery) {
+      return {
+        type: LanguageType.Natural,
+        confidence: 0,
+        reason: 'Empty query',
+        warnings: ['No input provided'],
+      };
+    }
+
+    const scores: Record<LanguageType, ScoreResult> = {
+      ppl: this._checkPpl(trimmedQuery),
+      keyvalue: this._checkKeyValue(trimmedQuery),
+      natural: this._checkNaturalLanguage(trimmedQuery),
+    };
+
+    const sorted = (Object.keys(scores) as Array<keyof typeof scores>).sort(
+      (a, b) => scores[b].score - scores[a].score
+    );
+    const best = sorted[0];
+    const fallback = scores[best].score < 0.4 ? LanguageType.Natural : best;
+
+    return {
+      type: fallback,
+      confidence: scores[best].score,
+      reason: scores[best].reason,
+      warnings: scores[best].warnings,
+    };
+  }
+
+  private _checkPpl(query: string): ScoreResult {
+    let score = 0;
+    const reasons: string[] = [];
+    const warnings: string[] = [];
+
+    if (this.pplStartPatterns.some((p) => p.test(query))) {
+      score += 0.5;
+      reasons.push('Starts with PPL pattern (e.g., source=, from, search)');
+    }
```

Comment:
The current regular expressions in `this.pplStartPatterns` could potentially be exploited for a ReDoS attack due to the use of multiple whitespace matchers (`\s*` and `\s+`). This can lead to catastrophic backtracking with carefully crafted malicious input.

A better approach would replace `^\s*` in those patterns with `^`, trim `query`, and limit its length to an arbitrary 50 chars.

```suggestion
    const startOfQuery = (query?.trim?.() || '').substring(0, 50).trim();
    if (this.pplStartPatterns.some((p) => p.test(startOfQuery))) {
      score += 0.5;
      reasons.push('Starts with PPL pattern (e.g., source=, from, search)');
    }
```